### ISSUE TITLE
Missing library prevents the Puzzle Editor from compiling maps

### ISSUE THREAD

**Issue by landswellsong** [ValveSoftware/steam-runtime#355]
When trying to compile the map in the Puzzle Editor, VBSP fails to start up:

```
VBSP failed.

ERROR: ld.so: object '/home/elfy/.local/share/Steam/ubuntu12_64/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
/home/elfy/.local/share/Steam/steamapps/common/Portal 2/bin/vbsp_linux: error while loading shared libraries: libuuid.so.1: cannot open shared object file: No such file or directory
```

The `libuuid.so.1` is indeed missing from the Portal 2's `bin` directory, so the operation must have relied on the host runtime, which is probably the case in ValveSoftware/portal2#235 as the reporter had a 64-bit host. The library in question can be found within steam's own runtime shipped under `~/.local/Steam`, copying or symlinking into Portal 2's `bin` directory remedies the problem.

Adding the library to the Portal 2's installation or making sure it uses the Steam runtime would fix this issue for new users.

System info just in case: https://gist.github.com/ohhai/e4bc5345be924f36c3241226795e5bc4

**Comment by kisak-valve** [comment]
Hello @landswellsong, Steam's LD_LIBRARY_PATH-based runtime should provide this library, or if you've disabled the runtime, then you (or the package maintainer who helped you disable the runtime) are responsible for resolving all missing dependencies.

It's not really clear how you've encountered this issue, but let's treat this as a steam runtime issue until there's a stronger indication the issue is elsewhere.

**Comment by landswellsong** [comment]
@kisak-valve right, but I did not specifically disable the runtime neither did the packager (there's an option for that), I'm fairly sure the game uses `LD_LIBRARY_PATH` one.

Evidence for my reasoning: 
1. `cat /proc/276713/environ  | tr '\0' '\n' | grep LIBRARY_PATH` : https://gist.github.com/landswellsong/c9581094d5d00e12d73c7d81e09bdfce
2. ` lsof -p 276713`: https://gist.github.com/landswellsong/b64b77dcfb5469022f8febd637352248

Hope that's helpful.

**Comment by smcv** [comment]
Breaking your `LD_LIBRARY_PATH` more legibly:

```
LD_LIBRARY_PATH=
/home/elfy/.local/share/Steam/steamapps/common/Portal 2/bin:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/pinned_libs_32:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/pinned_libs_64:
/usr/lib64/debiancompat:
/usr/lib/debiancompat:
/usr/lib64/fltk:
/usr/lib/fltk:
/usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0:
/usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/32:
/usr/lib64/itcl4.1.1:
/usr/lib64/itk4.1.0:
/lib64:
/usr/lib64:
/usr/local/lib64:
/lib:
/usr/lib:
/usr/local/lib:
/usr/lib/rust/lib:
/usr/lib/llvm/11/lib:
/usr/lib/llvm/11/lib64:
/usr/lib64/R/lib:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/lib/i386-linux-gnu:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/usr/lib/i386-linux-gnu:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/lib/x86_64-linux-gnu:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/usr/lib/x86_64-linux-gnu:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/lib:
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/usr/lib:
/home/elfy/.local/share/Steam/steamapps/common/Portal 2
```

**Comment by smcv** [comment]
Looks like that `portal2_linux` process *is* using the Steam Runtime (see the last few search-path entries).

Does VBSP perhaps "escape" from the Steam Runtime by resetting relevant environment variables, which would break this mechanism?

You could maybe find out by renaming `vbsp_linux` to `vbsp_linux.real`, and replacing it with a script like this:

```
#!/usr/bin/env bash
set -eu
me="$(readlink -f "$0")"
echo "vbsp_linux replacement: argv: $(printf '%q ' "$@")"
echo "vbsp_linux replacement: LDLP: ${LD_LIBRARY_PATH-}"
exec "$me.real" "$@"
```

> The libuuid.so.1 is indeed missing from the Portal 2's bin directory

It's not meant to be there, it's meant to be picked up from the `LD_LIBRARY_PATH` Steam Runtime. There should be a 32-bit copy in `/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/lib/i386-linux-gnu/libuuid.so.1` and a 64-bit copy in `/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/lib/x86_64-linux-gnu/libuuid.so.1`, both of which are in the search path, so they should be found without needing anything special.

It would be wrong to bundle another copy of `libuuid.so.1` with Portal 2, because that would break anything from the host system that is linked to `libuuid.so.1` and expects to see newer symbols that do not exist in the bundled copy.

**Comment by landswellsong** [comment]
@smcv 
>You could maybe find out by renaming vbsp_linux to vbsp_linux.real, and replacing it with a script like this:

That's interesting. The output is the following (modified to add linebreaks, but I re-checked with the raw value too):
```
vbsp_linux replacement: argv: -entity_limit 1750 -game /home/elfy/.local/share/Steam/steamapps/common/Portal\ 2/portal2/ /home/elfy/.local/share/Steam/steamapps/common/Portal\ 2/sdk_content/maps/preview.vmf 
vbsp_linux replacement: LDLP:
/home/elfy/.local/share/Steam/steamapps/common/Portal 2/bin/
/home/elfy/.local/share/Steam/steamapps/common/Portal 2/bin
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/pinned_libs_32
/home/elfy/.local/share/Steam/ubuntu12_32/steam-runtime/pinned

```
The last item seems truncated after the underscore, can it be some string processing issue that causes the tail of the variable to drop along with the relevant dirs which are at the end (given my snippet above)?

>It's not meant to be there, it's meant to be picked up from the LD_LIBRARY_PATH Steam Runtime.

Okay it makes sense and it's what I would have assumed. I just saw a bunch of libs in Portal's `bin` directory and thought it may be meant to be there.

**Comment by smcv** [comment]
> The last item seems truncated after the underscore

Indeed. Whatever is the parent process of `vbsp_linux` (Portal 2?) has prepended `/home/elfy/.local/share/Steam/steamapps/common/Portal 2/bin/` to the `LD_LIBRARY_PATH`, and seems to have truncated what's after that. This smells like a fixed-length buffer to me.

If the parent process is Portal 2, then that would point to this being a Portal 2 bug. @kisak-valve, perhaps you could bounce this back there?
